### PR TITLE
added lobster calc quality summary method to analyze module

### PR DESCRIPTION
**Pull request by naik-aakash**
#Changes

PR for issue #30

Added a convenience static method to ensure lobster calc reliablitly.  Checks implemented are similar to our Publication

- Check for minimal basis
- Charge spilling
- DOS comparisons summed and orbital wise Tanimoto
- Charge comparison with BVA

The output is like this at the moment. If you have any other suggestions @JaGeo please let me know.
Example

### CODE REVIEW

**Comment by JaGeo** [comment]
I would add a text output as well. Maybe similar to.our automatic-analysis. --json then creates a json

**Comment by naik-aakash** [comment]
> I would add a text output as well. Maybe similar to.our automatic-analysis. --json then creates a json

Yes, I intend to add this. Just wanted if anything more you think we should add besides these keys.

**Comment by naik-aakash** [comment]
- [ ] Add summary method to analyze
- [ ] Add text description
- [ ] Add tests

**Comment by naik-aakash** [comment]
Hi @JaGeo , the text descriptions are also generated and options to save the dict data as json is also added to cli. I have attached example outputs of text generated. DOS and BVA charge comparisons are optional as well.


Agree


Disagree

**Comment by naik-aakash** [comment]
### Updated text output
> The LOBSTER calculation used minimal basis. The absolute and total charge spilling for the calculation are 0.3 and 5.58, respectively. The atomic charge signs from Mulliken population analysis agree with Bond valence analysis. The atomic charge signs from Loewdin population analysis agree with Bond valence analysis. The Tanimoto index from DOS comparisons in energy range between -15, 0 eV for s, p, summed orbitals are : 0.274, 0.5281, 0.2737.

**Comment by naik-aakash** [comment]
Hi @JaGeo , have updated the tests here to increase the coverage of cli as well. I feel this PR is ready for review now 😄

File: lobsterpy/cohp/analyze.py [hunk 1]
```diff
@@ -869,3 +872,149 @@ def set_summary_dicts(self):
...
+            dos_lobster = doscar_lobster.completedos
+
+            vasprun = Vasprun(path_to_vasprun)
+            dos_vasp = vasprun.complete_dos
+
+            quality_dict["DOS_comparisons"] = {}  # type: ignore
+
+            for orb in dos_lobster.get_spd_dos():
+                fp_lobster_orb = dos_lobster.get_dos_fp(
+                    min_e=-15, max_e=0, n_bins=256, normalize=True, type=orb.name
```
[the remark this answers is not in the document]
Comment (JaGeo) [reply lobsterpy/cohp/analyze.py:992]: @naik-aakash  Any comment this part here?

File: lobsterpy/cohp/describe.py [hunk 2]
```diff
@@ -482,3 +482,66 @@ def write_description(self):
...
+                    )
+
+            elif key == "DOS_comparisons":
+                comp_types = []
+                tani_index = []
+                for orb in val:
+                    comp_types.append(orb.split("_")[-1])
+                    tani_index.append(str(val[orb]))
+                text_des.append(
+                    "The Tanimoto index from DOS comparisons in energy range between -15, 0 eV "
```
[the remark this answers is not in the document]
Comment (naik-aakash) [reply lobsterpy/cohp/describe.py:541]: Now at the moment it is fixed. But can reduce the default energy range and also make it possible to alter the energy range if one wants to

File: lobsterpy/cohp/analyze.py [hunk 3]
```diff
@@ -869,3 +872,149 @@ def set_summary_dicts(self):
...
+                else:
+                    quality_dict["Charges"]["BVA_Loewdin_agree"] = False  # type: ignore
+            except ValueError:
+                quality_dict["Charges"] = {}  # type: ignore
+                warnings.warn(
+                    "Oxidation states from BVA analyzer cannot be determined. "
+                    "Thus BVA charge comparison will be skipped"
+                )
+        if dos_comparison:
+            if "LSO" not in str(path_to_doscar).split("."):
```
[the remark this answers is not in the document]
Comment (naik-aakash) [reply lobsterpy/cohp/analyze.py:974]: Thanks! I overlooked this. It is really a good idea to also check for energy ranges. Will implement this

[the remark this answers is not in the document]
Comment (naik-aakash) [reply lobsterpy/cohp/analyze.py:976]: Added this check, if energy range does not exists, It reverts back to the default value of -5, 0 eV with a warning

File: lobsterpy/cohp/analyze.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (naik-aakash) [reply lobsterpy/cohp/analyze.py]: Added option to change energy range and number of bins. Will incorporate it to cli as well

File: lobsterpy/cohp/describe.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (naik-aakash) [reply lobsterpy/cohp/describe.py]: Hi @JaGeo , I now added possibility to change the energy range and number of bins to cli as well as method stores these information in the dict generated.

File: lobsterpy/cohp/analyze.py [hunk 4]
```diff
@@ -869,3 +872,149 @@ def set_summary_dicts(self):
...
+                lobs_charge = Charge(filename=path_to_charge)
+                for i in bond_valence.get_valences(structure=struct):
+                    if i >= 0:
+                        bva_oxi.append("POS")
+                    else:
+                        bva_oxi.append("NEG")
+
+                mull_oxi = []
+                for i in lobs_charge.Mulliken:
+                    if i >= 0:
```
[the remark this answers is not in the document]
Comment (naik-aakash) [reply lobsterpy/cohp/analyze.py:947]: I thought about it, but could not figure out any other way to check this. If you have any suggestions let me know

**Comment by naik-aakash** [comment]
# New updated outputs for calc quality dict and text

example_output_new

File: lobsterpy/test/test_cli.py [hunk 5]
```diff
@@ -237,6 +237,177 @@ def test_lobsterin_generation_error_userbasis(self, tmp_path):
...
+
+        sys.stdout = sys.__stdout__
+
+        ref_text = (
+            "The LOBSTER calculation used minimal basis. "
+            "The absolute and total charge spilling for the calculation are 0.3 and 5.58, respectively. "
+            "The projected wave function is completely orthonormalized as no bandOverlaps.lobster file is "
+            "generated during LOBSTER run. "
+            "The atomic charge signs from Mulliken population analysis agree with bond valence analysis. "
+            "The atomic charge signs from Loewdin population analysis agree with bond valence analysis. "
```
Comment (JaGeo) [inline lobsterpy/test/test_cli.py:269]: Could you add articles to the outputs? "the bond valence analysis"

File: lobsterpy/cohp/describe.py [hunk 6]
```diff
@@ -482,3 +482,98 @@ def write_description(self):
         """
         for textpart in self.text:
             print(textpart)
+
```
Comment (JaGeo) [inline lobsterpy/cohp/describe.py:485]: Shouldn't there be a test for this in test_describe rather than in cli?

**Comment by JaGeo** [comment]
@naik-aakash What happens if we only have very few points for the computation of the Tanimoto fingerprint or VASP and Lobster computations are too different?

File: lobsterpy/cli.py [hunk 7]
```diff
@@ -615,6 +704,95 @@ def run(args):
...
+        dos_comparison = args.doscomp
+
+        if dos_comparison:
+            dos_files = {
+                "doscar": "DOSCAR.lobster",
+                "vasprun": "vasprun.xml",
+            }
+
+            for arg, default_value in dos_files.items():
+                file_path = getattr(args, arg)
```
Comment (JaGeo) [inline lobsterpy/cli.py:764]: would be good to have some more descriptive names here.

File: lobsterpy/cohp/analyze.py [hunk 8]
```diff
@@ -869,3 +872,240 @@ def set_summary_dicts(self):
...
+
+            dos_lobster = doscar_lobster.completedos
+
+            vasprun = Vasprun(path_to_vasprun)
+            dos_vasp = vasprun.complete_dos
+
+            quality_dict["DOS_comparisons"] = {}  # type: ignore
+
+            for orb in dos_lobster.get_spd_dos():
+                if e_range[0] >= min(dos_vasp.energies) and e_range[0] >= min(
```
Comment (JaGeo) [inline lobsterpy/cohp/analyze.py:1028]: Your test cases cover all these ifs?

File: lobsterpy/cohp/describe.py [hunk 9]
```diff
@@ -482,3 +482,106 @@ def write_description(self):
...
+                if not val:
+                    text_des.append(
+                        "Consider rerunning the calculation with the minimum basis as well. Choosing a "
+                        "larger basis set is only recommended if you see a significant improvement of "
+                        "the charge spilling."
+                    )
+
+            elif key == "charge_spilling":
+                text_des.append(
+                    "The absolute and total charge spilling for the calculation are {} and {}, "
```
Comment (JaGeo) [inline lobsterpy/cohp/describe.py:509]: We should have probably directly added % here.

**Comment by naik-aakash** [comment]
> @naik-aakash What happens if we only have very few points for the computation of the Tanimoto fingerprint or VASP and Lobster computations are too different? I think we really need to test a few cases here before we merge.

I do have a warning printed out for cases when too less points are there in VASP and LOBSTER DOS computations <2000

File: lobsterpy/cohp/describe.py (hunk 9 above)
Comment (naik-aakash) [reply lobsterpy/cohp/describe.py:509]: Yeah missed out there % symbol here. Will add it

Comment (naik-aakash) [reply lobsterpy/cohp/describe.py]: Done

File: lobsterpy/cohp/analyze.py (hunk 8 above)
Comment (naik-aakash) [reply lobsterpy/cohp/analyze.py:1027]: Updated tests and now if covers all this 😃

**Comment by naik-aakash** [comment]
Hi @JaGeo , currently cli test will fail as the interface would need to be updated  and the `potcar-symbols` arg needs to be added

I will do this tomorrow, but using your suggestion of potcar symbol lists works fine

**Comment by naik-aakash** [comment]
> > # New updated outputs for calc quality dict and text
> > example_output_new
> 
> Thank you! Could you check the text again for missing articles, please? Also, "There exists" should better be "There are". Maybe copy the text into grammarly as your PI is also not the best person for English grammar.

I have also updated the texts to fix missing articles.

**Review by JaGeo — APPROVED** [review]

**Pull request merged**